### ISSUE TITLE
Update S2190: FN with try/catch block

### ISSUE THREAD

**Issue by jarvars** [SonarSource/sonar-dotnet#2342]
### Description

Sorry for my english, the example of the rule S2190 indicates the suggestion of a conditional to stop the method, ¿a try / catch block is thought for this even though it does not do something?

### Repro steps

Rule example:

```
int Pow(int num, int exponent)   // Noncompliant; no condition under which pow isn't re-called
{
  num = num * Pow(num, exponent-1);
  return num;  // this is never reached
}
```

Try/Catch example:
```
int Pow(int num, int exponent)   
        {
            try
            {
                num = num * Pow(num, exponent - 1);
                return num;  // this is never reached
            }
            catch (Exception ex) // Compliant;
            {
                registerLog(string.Concat(ex.Message, "\n ", ex.StackTrace), GetType(), Enumeration.Error);
                throw;
            }                      
        }
```

### Related information

* SonarC# Version: 7.3 (build 5690)instalado
* Visual Studio Version: Microsoft Visual Studio Professional 2017 - 15.9.8

**Comment by Evangelink** [comment]
Hi @jarvars,

You are right that this try/catch is not going to help with breaking the recursion so we should still raise an issue on this case. We currently don't because of some limitations in the way we are detecting and handling recursions.

Put in other words, yes the rule currently detects this case as valid even though it should not.

**Comment by jarvars** [comment]
@Evangelink thanks for the confirmation and details.

**Comment by andrei-epure-sonarsource** [comment]
as discussed with @abbas-sabra-sonarsource this FN can be fixed by adding finally clause in the instructions of the finally block to distinguish catch from finally

**Comment by andrei-epure-sonarsource** [comment]
Initially I though this was related to the FN introduced in  #2511 (which, as discussed with @abbas-sabra-sonarsource , can be fixed by adding finally clause in the instructions of the finally block to distinguish catch from finally), but it's not. 

Most likely the problem is due to the weird shape of the CFG which should be improved in #474

**Comment by andrei-epure-sonarsource** [comment]
due to the complexity of the necessary changes, we will probably need to do a sprint only on improving the CFG representation of the try. See #2550 for explanations for the try-catch part. This is related.
